File: crates/aero-audio/tests/audio_frame_clock.rs
use aero_audio::clock::AudioFrameClock;

const SAMPLE_RATE_HZ: u32 = 48_000;
const NS_PER_SEC: u64 = 1_000_000_000;

fn tick_60hz_ns(tick_index: u32) -> u64 {
    // 1 second / 60 = 16_666_666.666... ns. Use a common pattern that distributes the extra
    // 40 nanoseconds across the second: 40 ticks of 16_666_667ns and 20 ticks of 16_666_666ns.
    if tick_index < 40 {
        16_666_667
    } else {
        16_666_666
    }
}

#[test]
fn advance_one_second_is_exact() {
    let mut clock = AudioFrameClock::new(SAMPLE_RATE_HZ, 0);
    assert_eq!(clock.advance_to(NS_PER_SEC), SAMPLE_RATE_HZ as usize);
    assert_eq!(clock.frac_fp, 0);
}

#[test]
fn repeated_small_steps_sum_to_single_large_step() {
    let mut clock_single = AudioFrameClock::new(SAMPLE_RATE_HZ, 0);
    let single = clock_single.advance_to(NS_PER_SEC);

    let mut clock_steps = AudioFrameClock::new(SAMPLE_RATE_HZ, 0);
    let mut now_ns = 0u64;
    let mut total = 0usize;
    for tick in 0..60 {
        now_ns += tick_60hz_ns(tick);
        total += clock_steps.advance_to(now_ns);
    }

    assert_eq!(now_ns, NS_PER_SEC);
    assert_eq!(total, single);
    assert_eq!(total, SAMPLE_RATE_HZ as usize);
    assert_eq!(clock_steps.frac_fp, 0);
}

#[test]
fn no_drift_over_ten_minutes_at_60hz() {
    let mut clock = AudioFrameClock::new(SAMPLE_RATE_HZ, 0);
    let mut now_ns = 0u64;
    let mut total_frames = 0usize;

    for _second in 0..600u32 {
        for tick in 0..60u32 {
            now_ns += tick_60hz_ns(tick);
            total_frames += clock.advance_to(now_ns);
        }
    }

    assert_eq!(now_ns, NS_PER_SEC * 600);
    assert_eq!(total_frames, SAMPLE_RATE_HZ as usize * 600);
    assert_eq!(clock.frac_fp, 0);
}

#[test]
fn time_going_backwards_is_ignored() {
    let mut clock = AudioFrameClock::new(SAMPLE_RATE_HZ, 1_000);
    clock.advance_to(2_000);
    assert_eq!(clock.last_time_ns, 2_000);
    assert_eq!(clock.advance_to(1_500), 0);
    assert_eq!(clock.last_time_ns, 2_000);
}

File: web/src/audio/audio_frame_clock.ts
const NS_PER_SEC = 1_000_000_000n;

function toBigIntNs(nowNs: number | bigint): bigint {
  if (typeof nowNs === "bigint") return nowNs;
  if (!Number.isFinite(nowNs)) throw new Error(`nowNs must be a finite number, got ${String(nowNs)}`);
  if (nowNs < 0) throw new Error(`nowNs must be >= 0, got ${nowNs}`);
  // Converting an imprecise `number` into a `bigint` would make the clock's output
  // non-deterministic. Force callers to use `bigint` if they need larger ranges.
  if (nowNs > Number.MAX_SAFE_INTEGER) {
    throw new Error(`nowNs is too large to represent precisely as a number; pass a bigint instead (${nowNs})`);
  }
  return BigInt(Math.floor(nowNs));
}

/**
 * Convert a `performance.now()` timestamp (milliseconds) into an integer nanosecond
 * timestamp suitable for {@link AudioFrameClock}.
 *
 * Note: `performance.now()` returns a floating-point value with
 * implementation-defined precision (often microseconds, not true nanoseconds).
 * We use `Math.floor` to avoid rounding up due to floating-point error.
 */
export function perfNowMsToNs(perfNowMs: number): bigint {
  if (!Number.isFinite(perfNowMs)) throw new Error(`perfNowMs must be finite, got ${String(perfNowMs)}`);
  if (perfNowMs < 0) throw new Error(`perfNowMs must be >= 0, got ${perfNowMs}`);

  // Fast path: if `perfNowMs` is already an integer millisecond timestamp
  // (e.g. `Date.now()`), convert without going through floating-point
  // multiplication so large epoch millisecond values remain exact.
  if (Number.isSafeInteger(perfNowMs)) {
    return BigInt(perfNowMs) * 1_000_000n;
  }

  // Common case: `performance.now()` is relative to the page/worker time origin,
  // so `perfNowMs * 1e6` stays below `Number.MAX_SAFE_INTEGER` for typical
  // sessions (<~104 days). Use a single multiply+floor to avoid introducing extra
  // rounding error from splitting integer/fractional components.
  const maxMsForSafeMul = Number.MAX_SAFE_INTEGER / 1e6;
  if (perfNowMs <= maxMsForSafeMul) {
    return BigInt(Math.floor(perfNowMs * 1e6));
  }

  // Fallback: extremely long-running sessions with sub-ms precision. Convert the
  // integer milliseconds exactly, then add the fractional microseconds.
  const msInt = Math.floor(perfNowMs);
  const fracMs = perfNowMs - msInt;
  const fracUs = Math.floor(fracMs * 1e6);
  return BigInt(msInt) * 1_000_000n + BigInt(Math.max(0, Math.min(999_999, fracUs)));
}

/**
 * Best-effort monotonic nanosecond "now" for browser/worker scheduling.
 *
 * Prefers `performance.now()` (relative monotonic clock). Falls back to `Date.now()`
 * with millisecond precision if `performance` is unavailable.
 */
export function performanceNowNs(): bigint {
  const perfNow = globalThis.performance?.now;
  if (typeof perfNow === "function") return perfNowMsToNs(perfNow.call(globalThis.performance));
  // `Date.now()` is epoch milliseconds (integer), so we can convert without going
  // through floating-point.
  return BigInt(Date.now()) * 1_000_000n;
}

/**
 * Deterministic time→audio-frame conversion without cumulative rounding drift.
 *
 * Mirrors Rust `crates/aero-audio/src/clock.rs`.
 */
export class AudioFrameClock {
  readonly sampleRateHz: number;
  lastTimeNs: bigint;
  /**
   * Remainder accumulator in the same fixed-point units as Rust `frac_fp`.
   *
   * This is the remainder from dividing `delta_ns * sampleRateHz + fracNsTimesRate`
   * by `1_000_000_000` (nanoseconds per second). Always `< 1_000_000_000`.
   */
  fracNsTimesRate: bigint;

  readonly #sampleRateHzBig: bigint;

  constructor(sampleRateHz: number, startTimeNs: number | bigint) {
    if (!Number.isFinite(sampleRateHz)) {
      throw new Error(`sampleRateHz must be finite, got ${String(sampleRateHz)}`);
    }
    const sr = Math.floor(sampleRateHz);
    if (sr <= 0) throw new Error(`sampleRateHz must be > 0, got ${sampleRateHz}`);

    this.sampleRateHz = sr;
    this.#sampleRateHzBig = BigInt(sr);
    this.lastTimeNs = toBigIntNs(startTimeNs);
    this.fracNsTimesRate = 0n;
  }

  /**
   * Advance the clock to `nowNs` and return the number of audio frames that
   * elapsed since the previous call.
   *
   * If `nowNs` is earlier than the last observed time, the delta is treated as 0
   * and the internal time does not move backwards.
   */
  advanceTo(nowNs: number | bigint): number {
    const now = toBigIntNs(nowNs);
    if (now <= this.lastTimeNs) return 0;

    const deltaNs = now - this.lastTimeNs;
    this.lastTimeNs = now;

    const total = this.fracNsTimesRate + deltaNs * this.#sampleRateHzBig;
    const frames = total / NS_PER_SEC;
    this.fracNsTimesRate = total % NS_PER_SEC;

    const max = BigInt(Number.MAX_SAFE_INTEGER);
    return frames > max ? Number.MAX_SAFE_INTEGER : Number(frames);
  }
}
